### DISCUSSION TITLE
How to avoid "Emit skipped" error

### DISCUSSION THREAD

**Discussion by iamarpitpatidar** [TypeStrong/ts-node#1595] [Q&A] [answered]
### Short summary of the root cause:

"Emit skipped" error is thrown because a config file is written as `.ts` and resides *outside* of `rootDir`.

---

I am using commitlint which is using ts-node as a dep. Now somehow whenever I try to commit `ts-node` throws error `TypeScriptCompileError Emit Skipped` which is weird coz I compiled the project with tsc and it compiled succesfully.
I even tried it with just one file containing `console.log` please check the linked github repo

### Search Terms
Emit Skipped
commitlint

### Actual Behavior
Edit: just found that even after settings the `rootdir` in tsconfig `ts-node` still includes files outside from `src` directory, I think its a bug but I'm not sure.

### Expected Behavior
should not throw compile error and emit skipped.
Must not include files outside of `rootdir`

### Steps to reproduce the problem
```
yarn install
git add .
git commit -m "test: added a test"
```

### Minimal reproduction
https://github.com/iamarpitpatidar/ts-node-bug

### Specifications

* ts-node version: v10.4.0
* node version: v16.13.1
* TypeScript version: v4.5.4
* tsconfig.json, if you're using one:
```
{
  "compilerOptions": {
    "target": "ES2020",
    "module": "CommonJS",
    "outDir": "./dist",
    "rootDir": "./src",
    "moduleResolution": "Node",
    "lib": ["ESNext", "ESNext.AsyncIterable"],
    "esModuleInterop": true,
    "allowJs": true,
    "strict": true,
    "experimentalDecorators": true,
    "noEmitOnError": true,
    "baseUrl": ".",
    "types": ["@types/node"],
    "forceConsistentCasingInFileNames": true,
    "skipLibCheck": true
  },
  "include": ["src"],
  "exclude": ["node_modules", "dist"]
}
```
* Operating system and version: Zorin OS 16

**Comment by cspotcode** [comment]
I see you have a config file written in `.ts`, I assume something is trying to use ts-node to run it?  And it's not included in your tsconfig?

**Comment by iamarpitpatidar** [comment]
yeah, that's the thing the github I linked has the exact same config with just one file that I am using.. but wierd that it throws error,.

about `something is trying to use ts-node to run it` its commitlint as ofc the error only happens when husky run `commitlint` as pre-commit hook

@cspotcode do you think there is a way to fix this ?

**Comment by cspotcode** [comment, marked as answer]
ts-node uses typescript under the hood.  We need to get output JS to give to node to execute it.  TypeScript does not emit files that fall outside your rootDir.  So when ts-node is asked for the JS code from a file, and typescript doesn't give it to us, you get the error.

`transpileOnly` mode avoids this because we avoid typechecking complexity and use a simpler method of transformation.  Each file is transformed in isolation.  TypeScript doesn't care where the input file is located, and it always emits JS.

Or you can use `compilerOptions` overrides to modify your `rootDir` *only* when ts-node is running.

https://typestrong.org/ts-node/docs/configuration
https://typestrong.org/ts-node/docs/options

**Reply by cspotcode** [reply]
@iamarpitpatidar I've moved this to "Discussions" and marked this reply as the answer.  We can update this answer to help you and anyone else who may hit this error.

**Reply by iamarpitpatidar** [reply]
I see and do understand now why the error is happening, so I guess we can use `transpileOnly` as a temporary fix, but if we look for a solution so whose fault is it, I mean you said when ts-node is asked for JS code from a file, 
so, is that a `commitlint` bug as this lib is only calling ts-node on the config or I'm getting it wrong ?

another possible way (not ideal tho I guess) should I move the `tsconfig.json` to `rootdir`? again I don't think its ideal.

**Reply by cspotcode** [reply]
commitlint is doing the right thing:
It's invoking ts-node without changing any configuration, so it respects your ts-node settings in tsconfig.json.  This is good.  Some libraries mess with ts-node configuration, which is bad because it prevents users from configuring it for their projects.

ts-node is also doing the right thing:
It's typechecking, since that is the default behaviour, and asking TypeScript for emitted JS.  TypeScript doesn't give us anything.

We have a way to automatically avoid this error, but it has minor limitations: #1345  It would be a bit of a hack.  PRs welcome, if you feel like implementing this change!

Do you need typechecking of your commitlint config file?

* If yes: add `"ts-node": {"compilerOptions": {"rootDir": "."` override to your tsconfig
* If no: add `"ts-node": {"transpileOnly"` override to your tsconfig

**Reply by iamarpitpatidar** [reply]
ooh, I see okay now I get it. thanks mate @cspotcode been stuck there for a while.
btw if I were to use `transpileOnly` would that only be applicable to files outside the project or all files ?
